File: frontend/de/unika/ipd/grgen/ir/executable/MatchClassFilterFunction.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 7.1
 * Copyright (C) 2003-2025 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ir.executable;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import de.unika.ipd.grgen.ir.ContainedInPackage;
import de.unika.ipd.grgen.ir.Entity;
import de.unika.ipd.grgen.ir.Ident;
import de.unika.ipd.grgen.ir.Identifiable;
import de.unika.ipd.grgen.ir.type.DefinedMatchType;
import de.unika.ipd.grgen.ir.type.Type;

/**
 * Base type for match class filter functions (internal and external).
 */
public abstract class MatchClassFilterFunction extends Identifiable implements MatchClassFilter, ContainedInPackage
{
	private String packageContainedIn;

	/** A list of the parameters */
	protected List<Entity> params = new LinkedList<Entity>();

	/** A list of the parameter types, computed from the parameters */
	protected List<Type> parameterTypes = null;

	/** The match class we're a filter for */
	protected DefinedMatchType matchClass;

	public MatchClassFilterFunction(String name, Ident ident)
	{
		super(name, ident);
	}

	public void setMatchClass(DefinedMatchType matchClass)
	{
		this.matchClass = matchClass;
	}

	@Override
	public DefinedMatchType getMatchClass()
	{
		return matchClass;
	}

	@Override
	public String getPackageContainedIn()
	{
		return packageContainedIn;
	}

	public void setPackageContainedIn(String packageContainedIn)
	{
		this.packageContainedIn = packageContainedIn;
	}

	public String getFilterName()
	{
		return getIdent().toString();
	}

	/** Add a parameter to the match class filter function. */
	public void addParameter(Entity entity)
	{
		params.add(entity);
	}

	/** Get all parameters of this match class filter function. */
	public List<Entity> getParameters()
	{
		return Collections.unmodifiableList(params);
	}

	/** Get all parameter types of this match class filter function. */
	public List<Type> getParameterTypes()
	{
		if(parameterTypes == null) {
			parameterTypes = new LinkedList<Type>();
			for(Entity entity : getParameters()) {
				parameterTypes.add(entity.getType());
			}
		}
		return Collections.unmodifiableList(parameterTypes);
	}
}

File: frontend/de/unika/ipd/grgen/ir/executable/MatchClassFilterFunction.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>

namespace de.unika.ipd.grgen.ir.executable
{

using System.Collections.Generic;

using ContainedInPackage = de.unika.ipd.grgen.ir.ContainedInPackage;
using Entity = de.unika.ipd.grgen.ir.Entity;
using Ident = de.unika.ipd.grgen.ir.Ident;
using Identifiable = de.unika.ipd.grgen.ir.Identifiable;
using DefinedMatchType = de.unika.ipd.grgen.ir.type.DefinedMatchType;
using Type = de.unika.ipd.grgen.ir.type.Type;

/// <summary>
/// Base type for match class filter functions (internal and external).
/// </summary>
public abstract class MatchClassFilterFunction : Identifiable, MatchClassFilter, ContainedInPackage
{
	private string packageContainedIn;

	/// <summary>
	/// A list of the parameters </summary>
	protected internal List<Entity> @params = new List<Entity>();

	/// <summary>
	/// A list of the parameter types, computed from the parameters </summary>
	protected internal List<Type> parameterTypes = null;

	/// <summary>
	/// The match class we're a filter for </summary>
	protected internal DefinedMatchType matchClass;

	public MatchClassFilterFunction(string name, Ident ident)
		: base(name, ident)
	{
	}

	public virtual DefinedMatchType MatchClass
	{
		set
		{
			this.matchClass = value;
		}
		get
		{
			return matchClass;
		}
	}


	public virtual string PackageContainedIn
	{
		get
		{
			return packageContainedIn;
		}
		set
		{
			this.packageContainedIn = value;
		}
	}


	public virtual string FilterName
	{
		get
		{
			return Ident.ToString();
		}
	}

	/// <summary>
	/// Add a parameter to the match class filter function. </summary>
	public virtual void AddParameter(Entity entity)
	{
		@params.Add(entity);
	}

	/// <summary>
	/// Get all parameters of this match class filter function. </summary>
	public virtual IList<Entity> Parameters
	{
		get
		{
			return @params.AsReadOnly();
		}
	}

	/// <summary>
	/// Get all parameter types of this match class filter function. </summary>
	public virtual IList<Type> ParameterTypes
	{
		get
		{
			if(parameterTypes == null)
			{
				parameterTypes = new List<Type>();
				foreach(Entity entity in Parameters)
					parameterTypes.Add(entity.Type);
			}
			return parameterTypes.AsReadOnly();
		}
	}
}

}
